# Command Line Tools for Genomic Data Science

[Certificate](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/images/Coursera%20gencommand%202015.pdf)

#### Part of the Genomic Data Science Specialization

Introduces to the commands that you need to manage and analyze directories, files, and large sets of genomic data. This is the fourth course in the Genomic Big Data Science Specialization from Johns Hopkins University.

#### About the Course

This class will cover an introduction to command line tools for managing genomic data. It will start with basic Unix commands for managing files and directories, followed by a review of the main starndardized data formats and hands-on instruction on how to use popular tools such as Samtools, Bedtools, Bowtie, and application specific technologies such as Bcftools and Cufflinks. This course has a $99 value but being offered at a significant discount.

#### Course Syllabus
1. Unix commands

   -Basic commands
   
   -Pipes
   
2. Sequences and genomic features

   -Formats: Fastq, BED, GTF, SAM
   
   -Tools: Samtools, Bedtools
   
3. Alignment and sequence variation

   -Bowtie, BWA

   -Samtools, Bcftools
   
   -Application: variant calling

4. Transcriptomics

   -Tophat

   -Cufflinks
   
   -IGV
   
   -Application: RNA-seq analysis
   
#### Projects - Develop Pipeline for Next Generation Sequencing: Analyze DNA Alignment, DNA Assembly, Genetic Variants and RNA Sequencing in Linux

##### Project_A (Exam1)

Assume you sequenced and assembled the genome of Malus domestica (apple), and performed gene annotation. You then collected samples and ran RNA-seq experiments to determine sets of genes that are expressed in the various tissues. This information was stored, respectively, in the following files: “apple.genome”, “apple.genes”, “apple.condition{A,B,C}”.

NOTE: The apple genome and the apple gene annotations for this project were extracted from the Rosaceae Genome Database (RGD). Actual data have then been modified, and hence may not directly reflect the information in the original RGD records. Answer each question separately. Where multiple values are required, separate them by ‘,’ only, with no spaces. Do not use ‘,’ within numbers.

[Project_A Code](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam1/project_exam1.pdf)

[Project_A Solution](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam1/Exam1.pdf)

##### Project_B (Exam2)

As part of a larger project cataloging genetic variation in the plant Arabidopsis thaliana, you sequenced and assembled the genome of one strain (‘wu_0_A’), then mapped back the reads to the assembled genome. The resulting BAM file is included in the package ‘gencommand_proj2_data.tar.gz’. Using SAMtools and BEDtools as well as other Unix commands introduced in this course, examine the files and answer the following questions.

NOTE: Input data have been obtained and modified from those generated by the 1001 Genomes Project, accession ‘Wu_0_A’.

[Project_B Code](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam2/project_exam2.pdf)

[Project_B Solution](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam2/Exam2.pdf)

##### Project_C (Exam3)

As part of the effort to catalog genetic variation in the plant Arabidopsis thaliana, you resequenced the genome of one strain (‘wu_0_A’; genome file: ‘wu_0.v7.fas’), to determine genetic variants in this organism. The sequencing reads produced are in the file ‘wu_0_A_wgs.fastq’. Using the tools bowtie2, samtools and bcftools, develop a pipeline for variant calling in this genome. 

NOTE: Genome and re-sequencing data have been obtained and modified from those generated by the 1001 Genomes Project, accession ‘Wu_0_A’.

[Project_C Code](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam3/project_exam3.pdf)

[Project_C Solution](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam3/Exam3.pdf)

##### Project_D (Exam4)

You are performing an RNA-seq experiment to determine genes that are differentially expressed at different stages in the development of Arabidopsis thaliana shoot apical meristem. You collected samples at day 8 and day 16 (files “Day8.fastq” and “Day16.fastq”), extracted and sequenced the cellular mRNA, and are now set to perform the bioinformatics analysis. The reference genome you will need for the analysis is “athal_chr.fa” and the reference gene annotations are in “athal_genes.gtf”. Use default parameters unless otherwise specified. Sample command files that you can modify to create your own pipeline are provided in the file “commands.tar.gz”. All files are provided in the archive gencommand_proj4.tar.gz.

NOTE: The genome and annotation data were obtained and modified from the Arabidopsis Information Resources (TAIR) Database, and the RNA-seq reads were extracted from GenBank’s Short Read Archive (SRA).

[Project_D Code](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam4/project_exam4.pdf)

[Project_D Solution](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam4/Exam4.pdf)

[Commands](https://github.com/lanttern/Genomic-Data-Science/blob/master/Course4_Command%20Line%20Tools%20for%20Genomic%20Data%20Science/Project_Exam4/commands.pdf)
