---
title: "Exploring differential expression results"
output:
  html_document:
    toc: yes
    toc_float: yes
    toc_depth: 3
    highlight: pygments
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

[back to lesson's homepage](https://tavareshugo.github.io/data-carpentry-rnaseq/)

# Lesson Objectives

* Understand how MA plots can be useful to explore the results of differential 
gene expression tests
* Understand how scaling data can help in visualising gene expression trends

#### Further resources

* Book chapter from Holmes & Huber _Modern Statistics for Modern Biology_:
    * [High-Throughput Count Data](https://www.huber.embl.de/msmb/Chap-CountData.html)


# Setup

In your project's directory, create a new script called `04_gene_clustering.R`, 
and start with the following code:

```{r, message = FALSE}
##### setup ####

# load packages
library(tidyverse)

# read the data
trans_cts <- read_csv("./data/counts_transformed.csv")
sample_info <- read_csv("./data/sample_info.csv")
test_result <- read_csv("./data/test_result.csv")
```


# MA plots

Before starting with our gene clustering, it's useful to reduce the dimensionality of 
the data and work only with those genes that are likely to show some differences
of expression. 

We were given a table with the results of a differential analysis test between 
0 min and each of the other time-points:

```{r, echo = FALSE}
test_result
```


The `padj` column contains p-values (adjusted for [multiple testing](https://en.wikipedia.org/wiki/Multiple_comparisons_problem))
of each test's `comparison`. We can therefore use this table to help us 
focus on a subset of potentially interesting genes.

One way to visualise the differences between T0 and the other time points is with 
an [MA plot](https://en.wikipedia.org/wiki/MA_plot), which shows the average expression 
of each gene plotted against the log-fold-change in expression between the samples:

```{r, echo=FALSE, warning=FALSE}
test_result %>% 
  # add column which contains value only if padj < 0.01
  mutate(sig = ifelse(padj < 0.01, log2FoldChange, NA)) %>% 
  # make the plot
  ggplot(aes(baseMean, log2FoldChange)) +
  geom_point(alpha = 0.1) +
  geom_point(aes(y = sig), colour = "brown", size = 1) +
  scale_x_continuous(trans = "log10") +
  facet_wrap(vars(comparison))
```

----

**Exercise:**

> Try to create the _MA plot_ by yourself. 
> 
> Why is the fold-change reported on a log-scale?

[Link to full exercise](00_exercises.html#41_ma_plot)

----

The plot above is in agreement with our [PCA analysis](03_rnaseq_pca.html), 
which showed that cells from T30 were overall transcriptomically quite distinct 
from T0 cells.

From the coloured clouds of points, we can also see that more genes seem to 
markedly increase than decrease their expression in relation to T0.



# Visualising expression trends

Now, let's focus on those genes where `padj < 0.01` for any of the time-points 
and then see how their expression changes across time. 

We start by extracting the genes of interest based on our criteria:

```{r}
candidate_genes <- test_result %>% 
  filter(padj < 0.01) %>%    # filter table
  pull(gene) %>%             # extract the gene column as a vector
  unique()                   # retain only unique values
```

For convenience we also make a long version of our table (covered in the 
[exploratory analysis lesson](02_rnaseq_exploratory.html)):

```{r}
trans_cts_long <- trans_cts %>% 
  # convert to long format
  pivot_longer(cols = wt_0_r1:mut_180_r3, names_to = "sample", values_to = "cts")  %>% 
  # join with sample info table
  full_join(sample_info, by = ("sample"))
```

Now we filter our table to retain only genes of interest and summarise our 
counts per time point:

```{r}
trans_cts_mean <- trans_cts_long %>% 
  # filter genes of interest
  filter(gene %in% candidate_genes) %>% 
  # for each gene, strain and minute
  group_by(gene, strain, minute) %>% 
  # calculate mean and number of replicates
  summarise(mean_cts = mean(cts),
            nrep = n()) %>% 
  # remove grouping from downstream analysis
  ungroup()

head(trans_cts_mean)
```

We want to explore _patterns_ of expression across time in the two strains. 
We can plot each gene's expression across time (although the result is very confusing!)

```{r}
trans_cts_mean %>% 
  ggplot(aes(minute, mean_cts)) +
  geom_line(aes(group = gene), alpha = 0.3) +
  facet_grid(rows = vars(strain))
```


## Scaling data to improve visualisation

When we are interested in _relative_ patterns of expression, it is very useful to 
transform our data to ensure that the different genes are represented on a comparable scale.

You could imagine that two genes might have changed by the same magnitude (say, 
doubled the expression between time points), but their base mean levels of expression 
might have been quite different. For example, one gene might have changed from 
10 to 20 and another from 100 to 200.

If what we’re interested in is the _relative change_ in expression, then those 
two genes will appear more different than they really are in our plots.

A useful data transformation in this case is to center and scale each genes’ 
expression by their mean and standard deviation, respectively. The values thus 
obtained are known as [z-scores](https://en.wikipedia.org/wiki/Standard_score), 
and can be interpreted as the “number of standard deviations away from the mean”. 
A positive z-score means that the gene’s expression was above the average across 
samples, whereas a negative one means it was below average. 
A value of zero means the gene’s expression was exactly average.

Let's re-do our data summary, but scale the each gene's expression data first:

```{r}
trans_cts_mean <- trans_cts_long %>% 
  # filter to retain only genes of interest
  filter(gene %in% candidate_genes) %>% 
  # for each gene
  group_by(gene) %>% 
  # scale the cts column
  mutate(cts_scaled = (cts - mean(cts))/sd(cts)) %>% 
  # for each gene, strain and minute
  group_by(gene, strain, minute) %>%
  # calculate the mean (scaled) cts
  summarise(mean_cts_scaled = mean(cts_scaled),
            nrep = n()) %>% 
  ungroup()
```

And now we can re-do the plot, and see how this transformation changes our 
perception of the data:

```{r}
trans_cts_mean %>%
  ggplot(aes(minute, mean_cts_scaled)) +
  geom_line(aes(group = gene), alpha = 0.2) + 
  geom_hline(yintercept = 0, colour = "brown", linetype = "dashed") +
  facet_grid(rows = vars(strain))
```

As you see, this is a substantially clearer picture than before, we can clearly 
see some patterns there. It's not perfect yet: as you can see there are different 
"populations" of genes with different trends. 
In the [clustering lesson](04b_rnaseq_clustering.html) we will see how we can partition 
the genes into different categories and separate them out so we don't have this 
terrible [spaghetti plot](https://www.data-to-viz.com/caveat/spaghetti.html).


